From: TalkOps <noreply@github.com>
Subject: [PATCH] Backport PCRE2 support from FreeSWITCH master to v1.10.12

Debian 13 (trixie) no longer ships PCRE1 (libpcre3). FreeSWITCH master
migrated to PCRE2, but no release contains it yet. This patch is the
combined diff of these upstream commits, limited to configure.ac,
Makefile.am and the sources of modules TalkOps builds:

  65bc7c1 [Core] Upgrade FreeSWITCH to use PCRE2 library (signalwire/freeswitch#2845)
  9092470 [Core] Define SWITCH_PCRE2 macro (signalwire/freeswitch#2858)
  02549c1 [mod_dialplan_xml] Fix double free after upgrade to pcre2 (#2946)

Drop this patch when upgrading to a FreeSWITCH release that includes them.
---
diff --git a/Makefile.am b/Makefile.am
index fdd339a..f40e24b 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -231,9 +231,9 @@ CORE_LIBS+=libfreeswitch_libyuv.la
 endif
 
 lib_LTLIBRARIES	         = libfreeswitch.la
-libfreeswitch_la_CFLAGS  = $(CORE_CFLAGS) $(SQLITE_CFLAGS) $(GUMBO_CFLAGS) $(FVAD_CFLAGS) $(FREETYPE_CFLAGS) $(CURL_CFLAGS) $(PCRE_CFLAGS) $(SPEEX_CFLAGS) $(LIBEDIT_CFLAGS) $(openssl_CFLAGS) $(SOFIA_SIP_CFLAGS) $(AM_CFLAGS) $(TPL_CFLAGS)
+libfreeswitch_la_CFLAGS  = $(CORE_CFLAGS) $(SQLITE_CFLAGS) $(GUMBO_CFLAGS) $(FVAD_CFLAGS) $(FREETYPE_CFLAGS) $(CURL_CFLAGS) $(PCRE2_CFLAGS) $(SPEEX_CFLAGS) $(LIBEDIT_CFLAGS) $(openssl_CFLAGS) $(SOFIA_SIP_CFLAGS) $(AM_CFLAGS) $(TPL_CFLAGS)
 libfreeswitch_la_LDFLAGS = -version-info 1:0:0 $(AM_LDFLAGS) $(PLATFORM_CORE_LDFLAGS) -no-undefined
-libfreeswitch_la_LIBADD  = $(CORE_LIBS) $(APR_LIBS) $(SQLITE_LIBS) $(GUMBO_LIBS) $(FVAD_LIBS) $(FREETYPE_LIBS) $(CURL_LIBS) $(PCRE_LIBS) $(SPEEX_LIBS) $(LIBEDIT_LIBS) $(SYSTEMD_LIBS) $(openssl_LIBS) $(PLATFORM_CORE_LIBS) $(TPL_LIBS) $(SPANDSP_LIBS) $(SOFIA_SIP_LIBS)
+libfreeswitch_la_LIBADD  = $(CORE_LIBS) $(APR_LIBS) $(SQLITE_LIBS) $(GUMBO_LIBS) $(FVAD_LIBS) $(FREETYPE_LIBS) $(CURL_LIBS) $(PCRE2_LIBS) $(SPEEX_LIBS) $(LIBEDIT_LIBS) $(SYSTEMD_LIBS) $(openssl_LIBS) $(PLATFORM_CORE_LIBS) $(TPL_LIBS) $(SPANDSP_LIBS) $(SOFIA_SIP_LIBS)
 libfreeswitch_la_DEPENDENCIES = $(BUILT_SOURCES)
 
 if HAVE_PNG
diff --git a/configure.ac b/configure.ac
index 469d539..0daefca 100644
--- a/configure.ac
+++ b/configure.ac
@@ -1323,7 +1323,7 @@ PKG_CHECK_MODULES([TPL], [libtpl >= 1.5],[
 
 PKG_CHECK_MODULES([SQLITE], [sqlite3 >= 3.6.20])
 PKG_CHECK_MODULES([CURL], [libcurl >= 7.19])
-PKG_CHECK_MODULES([PCRE], [libpcre >= 7.8])
+PKG_CHECK_MODULES([PCRE2], [libpcre2-8 >=  10.00])
 PKG_CHECK_MODULES([SPEEX], [speex >= 1.2rc1 speexdsp >= 1.2rc1])
 PKG_CHECK_MODULES([YAML], [yaml-0.1 >= 0.1.4],[
   AM_CONDITIONAL([HAVE_YAML],[true])],[
diff --git a/src/include/switch.h b/src/include/switch.h
index 68c2004..3ac1176 100644
--- a/src/include/switch.h
+++ b/src/include/switch.h
@@ -172,7 +172,7 @@
  *		- APR (http://apr.apache.org)
  *		- APR-Util (http://apr.apache.org)
  *		- SQLite (http://www.sqlite.org)
- *		- Pcre (http://www.pcre.org/)
+ *		- Pcre2 (http://www.pcre.org/)
  *		- SRTP (http://srtp.sourceforge.net/srtp.html)
  *
  *	Additionally, the various external modules make use of several external modules:
diff --git a/src/include/switch_regex.h b/src/include/switch_regex.h
index 22147c1..f00f8b4 100644
--- a/src/include/switch_regex.h
+++ b/src/include/switch_regex.h
@@ -25,7 +25,7 @@
  *
  * Michael Jerris <mike@jerris.com>
  *
- * switch_regex.h -- pcre wrapper and extensions Header
+ * switch_regex.h -- pcre2 wrapper and extensions Header
  *
  */
 /*! \file switch_regex.h
@@ -40,18 +40,26 @@ SWITCH_BEGIN_EXTERN_C
  * @ingroup FREESWITCH
  * @{
  */
-	typedef struct real_pcre switch_regex_t;
 
-SWITCH_DECLARE(switch_regex_t *) switch_regex_compile(const char *pattern, int options, const char **errorptr, int *erroroffset,
-													  const unsigned char *tables);
+#define SWITCH_PCRE2 1
 
-SWITCH_DECLARE(int) switch_regex_copy_substring(const char *subject, int *ovector, int stringcount, int stringnumber, char *buffer, int size);
+typedef struct pcre2_real_code switch_regex_t;
+typedef struct pcre2_real_match_data_8 switch_regex_match_t;
+typedef struct pcre2_real_compile_context_8 switch_regex_compile_context_t;
 
+SWITCH_DECLARE(switch_regex_t *) switch_regex_compile(const char *pattern, int options, int *errorcode, unsigned int *erroroffset,
+													  switch_regex_compile_context_t *ccontext);
+
+SWITCH_DECLARE(int) switch_regex_copy_substring(switch_regex_match_t *match_data, int stringnumber, char *buffer, size_t *size);
+
+SWITCH_DECLARE(void) switch_regex_match_free(void *data);
 SWITCH_DECLARE(void) switch_regex_free(void *data);
 
-SWITCH_DECLARE(int) switch_regex_perform(const char *field, const char *expression, switch_regex_t **new_re, int *ovector, uint32_t olen);
-SWITCH_DECLARE(void) switch_perform_substitution(switch_regex_t *re, int match_count, const char *data, const char *field_data,
-												 char *substituted, switch_size_t len, int *ovector);
+SWITCH_DECLARE(int) switch_regex_perform(const char *field, const char *expression, switch_regex_t **new_re, switch_regex_match_t **new_match_data);
+#define switch_regex(field, expression) switch_regex_perform(field, expression, NULL, NULL)
+
+SWITCH_DECLARE(void) switch_perform_substitution(switch_regex_match_t *match_data, const char *data,
+												 char *substituted, switch_size_t len);
 
 /*!
  \brief Function to evaluate an expression against a string
@@ -70,12 +78,17 @@ SWITCH_DECLARE(switch_status_t) switch_regex_match(const char *target, const cha
 */
 SWITCH_DECLARE(switch_status_t) switch_regex_match_partial(const char *target, const char *expression, int *partial_match);
 
-SWITCH_DECLARE(void) switch_capture_regex(switch_regex_t *re, int match_count, const char *field_data,
-										  int *ovector, const char *var, switch_cap_callback_t callback, void *user_data);
+SWITCH_DECLARE(void) switch_capture_regex(switch_regex_match_t *match_data, int match_count,
+										  const char *var, switch_cap_callback_t callback, void *user_data);
 
 SWITCH_DECLARE_NONSTD(void) switch_regex_set_var_callback(const char *var, const char *val, void *user_data);
 SWITCH_DECLARE_NONSTD(void) switch_regex_set_event_header_callback(const char *var, const char *val, void *user_data);
 
+#define switch_regex_match_safe_free(match_data)	if (match_data) {\
+				switch_regex_match_free(match_data);\
+				match_data = NULL;\
+			}
+
 #define switch_regex_safe_free(re)	if (re) {\
 				switch_regex_free(re);\
 				re = NULL;\
diff --git a/src/mod/applications/mod_commands/mod_commands.c b/src/mod/applications/mod_commands/mod_commands.c
index 5b9620a..970fa35 100644
--- a/src/mod/applications/mod_commands/mod_commands.c
+++ b/src/mod/applications/mod_commands/mod_commands.c
@@ -2014,7 +2014,7 @@ SWITCH_STANDARD_API(replace_function)
 SWITCH_STANDARD_API(regex_function)
 {
 	switch_regex_t *re = NULL;
-	int ovector[30];
+	switch_regex_match_t *match_data = NULL;
 	int argc;
 	char *mydata = NULL, *argv[4];
 	size_t len = 0;
@@ -2054,7 +2054,7 @@ SWITCH_STANDARD_API(regex_function)
 		goto error;
 	}
 
-	proceed = switch_regex_perform(argv[0], argv[1], &re, ovector, sizeof(ovector) / sizeof(ovector[0]));
+	proceed = switch_regex_perform(argv[0], argv[1], &re, &match_data);
 
 	if (argc > 2) {
 		char *flags = "";
@@ -2069,7 +2069,7 @@ SWITCH_STANDARD_API(regex_function)
 			switch_assert(substituted);
 			memset(substituted, 0, len);
 			switch_replace_char(argv[2], '%', '$', SWITCH_FALSE);
-			switch_perform_substitution(re, proceed, argv[2], argv[0], substituted, len, ovector);
+			switch_perform_substitution(match_data, argv[2], substituted, len);
 
 			stream->write_function(stream, "%s", substituted);
 			free(substituted);
@@ -2091,6 +2091,7 @@ SWITCH_STANDARD_API(regex_function)
   error:
 	stream->write_function(stream, "-ERR");
   ok:
+	switch_regex_match_safe_free(match_data);
 	switch_regex_safe_free(re);
 	switch_safe_free(mydata);
 	return SWITCH_STATUS_SUCCESS;
diff --git a/src/mod/applications/mod_dptools/mod_dptools.c b/src/mod/applications/mod_dptools/mod_dptools.c
index 58fce5f..431088f 100644
--- a/src/mod/applications/mod_dptools/mod_dptools.c
+++ b/src/mod/applications/mod_dptools/mod_dptools.c
@@ -3211,15 +3211,17 @@ SWITCH_STANDARD_APP(capture_function)
 {
 	char *argv[3] = { 0 };
 	switch_regex_t *re = NULL;
-	int ovector[30] = {0};
+	switch_regex_match_t *match_data = NULL;
 	char *lbuf;
 	int proceed;
 
 	if (!zstr(data) && (lbuf = switch_core_session_strdup(session, data))
 		&& switch_separate_string(lbuf, '|', argv, (sizeof(argv) / sizeof(argv[0]))) == 3) {
-		if ((proceed = switch_regex_perform(argv[1], argv[2], &re, ovector, sizeof(ovector) / sizeof(ovector[0])))) {
-			switch_capture_regex(re, proceed, argv[1], ovector, argv[0], switch_regex_set_var_callback, session);
+		if ((proceed = switch_regex_perform(argv[1], argv[2], &re, &match_data))) {
+			switch_capture_regex(match_data, proceed, argv[0], switch_regex_set_var_callback, session);
 		}
+
+		switch_regex_match_safe_free(match_data);
 		switch_regex_safe_free(re);
 	} else {
 		switch_log_printf(SWITCH_CHANNEL_SESSION_LOG(session), SWITCH_LOG_ERROR, "No data specified.\n");
diff --git a/src/mod/applications/mod_enum/mod_enum.c b/src/mod/applications/mod_enum/mod_enum.c
index aa87f24..3a6efba 100644
--- a/src/mod/applications/mod_enum/mod_enum.c
+++ b/src/mod/applications/mod_enum/mod_enum.c
@@ -365,7 +365,8 @@ static void parse_naptr(const ldns_rr *naptr, const char *number, enum_record_t
 
 	if (service && regex && replace) {
 		switch_regex_t *re = NULL, *re2 = NULL;
-		int proceed = 0, ovector[30];
+		switch_regex_match_t *match_data = NULL, *match_data2 = NULL;
+		int proceed = 0;
 		char *substituted = NULL;
 		char *substituted_2 = NULL;
 		char *orig_uri;
@@ -374,17 +375,18 @@ static void parse_naptr(const ldns_rr *naptr, const char *number, enum_record_t
 		int supported = 0;
 		uint32_t len = 0;
 
-		if ((proceed = switch_regex_perform(number, regex, &re, ovector, sizeof(ovector) / sizeof(ovector[0])))) {
+		if ((proceed = switch_regex_perform(number, regex, &re, &match_data))) {
 			if (strchr(regex, '(')) {
 				len = (uint32_t) (strlen(number) + strlen(replace) + 10) * proceed;
 				if (!(substituted = malloc(len))) {
 					switch_log_printf(SWITCH_CHANNEL_LOG, SWITCH_LOG_CRIT, "Memory Error!\n");
+					switch_regex_match_safe_free(match_data);
 					switch_regex_safe_free(re);
 					goto end;
 				}
 				memset(substituted, 0, len);
 
-				switch_perform_substitution(re, proceed, replace, number, substituted, len, ovector);
+				switch_perform_substitution(match_data, replace, substituted, len);
 				orig_uri = substituted;
 			} else {
 				orig_uri = replace;
@@ -398,7 +400,7 @@ static void parse_naptr(const ldns_rr *naptr, const char *number, enum_record_t
 					continue;
 				}
 
-				if ((proceed = switch_regex_perform(uri, route->regex, &re2, ovector, sizeof(ovector) / sizeof(ovector[0])))) {
+				if ((proceed = switch_regex_perform(uri, route->regex, &re2, &match_data2))) {
 					switch_event_t *event = NULL;
 
 					if (strchr(route->regex, '(')) {
@@ -406,6 +408,8 @@ static void parse_naptr(const ldns_rr *naptr, const char *number, enum_record_t
 						if (!(substituted_2 = malloc(len))) {
 							switch_log_printf(SWITCH_CHANNEL_LOG, SWITCH_LOG_CRIT, "Memory Error!\n");
 							switch_safe_free(substituted);
+							switch_regex_match_safe_free(match_data);
+							switch_regex_match_safe_free(match_data2);
 							switch_regex_safe_free(re);
 							switch_regex_safe_free(re2);
 							switch_mutex_unlock(MUTEX);
@@ -413,7 +417,7 @@ static void parse_naptr(const ldns_rr *naptr, const char *number, enum_record_t
 						}
 						memset(substituted_2, 0, len);
 
-						switch_perform_substitution(re2, proceed, route->replace, uri, substituted_2, len, ovector);
+						switch_perform_substitution(match_data2, route->replace, substituted_2, len);
 						uri = substituted_2;
 					} else {
 						uri = route->replace;
@@ -434,6 +438,7 @@ static void parse_naptr(const ldns_rr *naptr, const char *number, enum_record_t
 				}
 				switch_safe_free(uri_expanded);
 				switch_safe_free(substituted_2);
+				switch_regex_match_safe_free(match_data2);
 				switch_regex_safe_free(re2);
 			}
 			switch_mutex_unlock(MUTEX);
@@ -443,6 +448,7 @@ static void parse_naptr(const ldns_rr *naptr, const char *number, enum_record_t
 			}
 
 			switch_safe_free(substituted);
+			switch_regex_match_safe_free(match_data);
 			switch_regex_safe_free(re);
 		}
 	}
diff --git a/src/mod/applications/mod_lcr/mod_lcr.c b/src/mod/applications/mod_lcr/mod_lcr.c
index ab7c163..3525d46 100644
--- a/src/mod/applications/mod_lcr/mod_lcr.c
+++ b/src/mod/applications/mod_lcr/mod_lcr.c
@@ -166,7 +166,8 @@ static void lcr_destroy(lcr_route route)
 static const char *do_cid(switch_memory_pool_t *pool, const char *cid, const char *number, switch_core_session_t *session)
 {
 	switch_regex_t *re = NULL;
-	int proceed = 0, ovector[30];
+	switch_regex_match_t *match_data = NULL;
+	int proceed = 0;
 	char *substituted = NULL;
 	uint32_t len = 0;
 	char *src = NULL;
@@ -230,23 +231,25 @@ static const char *do_cid(switch_memory_pool_t *pool, const char *cid, const cha
 		switch_log_printf(SWITCH_CHANNEL_SESSION_LOG(session), SWITCH_LOG_DEBUG, "expanded src: %s, dst: %s\n", src, dst);
 	}
 
-	if ((proceed = switch_regex_perform(number, src, &re, ovector, sizeof(ovector) / sizeof(ovector[0])))) {
+	if ((proceed = switch_regex_perform(number, src, &re, &match_data))) {
 		len = (uint32_t) (strlen(src) + strlen(dst) + 10) * proceed; /* guestimate size */
 		if (!(substituted = switch_core_alloc(pool, len))) {
 			switch_log_printf(SWITCH_CHANNEL_SESSION_LOG(session), SWITCH_LOG_ERROR, "Memory Error!\n");
 			goto done;
 		}
 		memset(substituted, 0, len);
-		switch_perform_substitution(re, proceed, dst, number, substituted, len, ovector);
+		switch_perform_substitution(match_data, dst, substituted, len);
 	} else {
 		goto done;
 	}
 
+	switch_regex_match_safe_free(match_data);
 	switch_regex_safe_free(re);
 
 	return substituted;
 
 done:
+	switch_regex_match_safe_free(match_data);
 	switch_regex_safe_free(re);
 	switch_safe_free(tmp_regex);
 	return number;
diff --git a/src/mod/applications/mod_sms/mod_sms.c b/src/mod/applications/mod_sms/mod_sms.c
index 68ad7d7..5375308 100644
--- a/src/mod/applications/mod_sms/mod_sms.c
+++ b/src/mod/applications/mod_sms/mod_sms.c
@@ -124,6 +124,7 @@ static int parse_exten(switch_event_t *event, switch_xml_t xexten, switch_event_
 	int proceed = 0;
 	char *expression_expanded = NULL, *field_expanded = NULL;
 	switch_regex_t *re = NULL;
+	switch_regex_match_t *match_data = NULL;
 	const char *to = switch_event_get_header(event, "to");
 	const char *tzoff = NULL, *tzname_ = NULL;
 	int offset = 0;
@@ -143,7 +144,6 @@ static int parse_exten(switch_event_t *event, switch_xml_t xexten, switch_event_
 		char *do_break_a = NULL;
 		char *expression = NULL;
 		const char *field_data = NULL;
-		int ovector[30];
 		switch_bool_t anti_action = SWITCH_TRUE;
 		break_t do_break_i = BREAK_ON_FALSE;
 		int time_match;
@@ -214,7 +214,7 @@ static int parse_exten(switch_event_t *event, switch_xml_t xexten, switch_event_
 				field_data = "";
 			}
 
-			if ((proceed = switch_regex_perform(field_data, expression, &re, ovector, sizeof(ovector) / sizeof(ovector[0])))) {
+			if ((proceed = switch_regex_perform(field_data, expression, &re, &match_data))) {
 				switch_log_printf(SWITCH_CHANNEL_LOG_CLEAN, SWITCH_LOG_DEBUG,
 								  "Chatplan: %s Regex (PASS) [%s] %s(%s) =~ /%s/ break=%s\n",
 								  to, exten_name, field, field_data, expression, do_break_a ? do_break_a : "on-false");
@@ -271,7 +271,7 @@ static int parse_exten(switch_event_t *event, switch_xml_t xexten, switch_event_
 		} else {
 			if (field && strchr(expression, '(')) {
 				switch_event_add_header_string(event, SWITCH_STACK_BOTTOM, "DP_MATCH", NULL);
-				switch_capture_regex(re, proceed, field_data, ovector, "DP_MATCH", switch_regex_set_event_header_callback, event);
+				switch_capture_regex(match_data, proceed, "DP_MATCH", switch_regex_set_event_header_callback, event);
 			}
 
 			for (xaction = switch_xml_child(xcond, "action"); xaction; xaction = xaction->next) {
@@ -297,7 +297,7 @@ static int parse_exten(switch_event_t *event, switch_xml_t xexten, switch_event_
 						abort();
 					}
 					memset(substituted, 0, len);
-					switch_perform_substitution(re, proceed, data, field_data, substituted, len, ovector);
+					switch_perform_substitution(match_data, data, substituted, len);
 					app_data = substituted;
 				} else {
 					app_data = data;
@@ -326,6 +326,8 @@ static int parse_exten(switch_event_t *event, switch_xml_t xexten, switch_event_
 				switch_safe_free(substituted);
 			}
 		}
+
+		switch_regex_match_safe_free(match_data);
 		switch_regex_safe_free(re);
 
 		if (((anti_action == SWITCH_FALSE && do_break_i == BREAK_ON_TRUE) ||
@@ -335,6 +337,7 @@ static int parse_exten(switch_event_t *event, switch_xml_t xexten, switch_event_
 	}
 
   done:
+	switch_regex_match_safe_free(match_data);
 	switch_regex_safe_free(re);
 	switch_safe_free(field_expanded);
 	switch_safe_free(expression_expanded);
diff --git a/src/mod/applications/mod_translate/mod_translate.c b/src/mod/applications/mod_translate/mod_translate.c
index 34b1ad5..cc31b1f 100644
--- a/src/mod/applications/mod_translate/mod_translate.c
+++ b/src/mod/applications/mod_translate/mod_translate.c
@@ -117,7 +117,8 @@ static void translate_number(char *number, char *profile, char **translated, swi
 	translate_rule_t *hi = NULL;
 	translate_rule_t *rule = NULL;
 	switch_regex_t *re = NULL;
-	int proceed = 0, ovector[30];
+	switch_regex_match_t *match_data = NULL;
+	int proceed = 0;
 	char *substituted = NULL, *subbed = NULL;
 	uint32_t len = 0;
 
@@ -136,17 +137,18 @@ static void translate_number(char *number, char *profile, char **translated, swi
 
 	for (rule = hi; rule; rule = rule->next) {
 		switch_log_printf(SWITCH_CHANNEL_LOG, SWITCH_LOG_DEBUG, "%s =~ /%s/\n", number, rule->regex);
-		if ((proceed = switch_regex_perform(number, rule->regex, &re, ovector, sizeof(ovector) / sizeof(ovector[0])))) {
+		if ((proceed = switch_regex_perform(number, rule->regex, &re, &match_data))) {
 			len = (uint32_t) (strlen(number) + strlen(rule->replace) + 10) * proceed;
 			if (!(substituted = malloc(len))) {
 				switch_log_printf(SWITCH_CHANNEL_LOG, SWITCH_LOG_CRIT, "Memory Error!\n");
+				switch_regex_match_safe_free(match_data);
 				switch_regex_safe_free(re);
 				goto end;
 			}
 
 			memset(substituted, 0, len);
 
-			switch_perform_substitution(re, proceed, rule->replace, number, substituted, len, ovector);
+			switch_perform_substitution(match_data, rule->replace, substituted, len);
 
 			if ((switch_string_var_check_const(substituted) || switch_string_has_escaped_data(substituted))) {
 				switch_log_printf(SWITCH_CHANNEL_LOG, SWITCH_LOG_DEBUG, "perform variable expansion\n");
@@ -169,6 +171,7 @@ static void translate_number(char *number, char *profile, char **translated, swi
 				switch_safe_free(subbed);
 			}
 
+			switch_regex_match_safe_free(match_data);
 			switch_regex_safe_free(re);
 			break;
 		}
diff --git a/src/mod/dialplans/mod_dialplan_asterisk/mod_dialplan_asterisk.c b/src/mod/dialplans/mod_dialplan_asterisk/mod_dialplan_asterisk.c
index 652d292..f51f78d 100644
--- a/src/mod/dialplans/mod_dialplan_asterisk/mod_dialplan_asterisk.c
+++ b/src/mod/dialplans/mod_dialplan_asterisk/mod_dialplan_asterisk.c
@@ -170,9 +170,8 @@ SWITCH_STANDARD_DIALPLAN(asterisk_dialplan_hunt)
 				char *expression = NULL, expression_buf[1024] = { 0 };
 				char substituted[2048] = "";
 				const char *field_data = caller_profile->destination_number;
-				int proceed = 0;
 				switch_regex_t *re = NULL;
-				int ovector[30] = { 0 };
+				switch_regex_match_t *match_data = NULL;
 				char *cid = NULL;
 
 				expression = expression_buf;
@@ -221,7 +220,8 @@ SWITCH_STANDARD_DIALPLAN(asterisk_dialplan_hunt)
 						field_data = "";
 					}
 
-					if (!(proceed = switch_regex_perform(field_data, expression, &re, ovector, sizeof(ovector) / sizeof(ovector[0])))) {
+					if (!(switch_regex_perform(field_data, expression, &re, &match_data))) {
+						switch_regex_match_safe_free(match_data);
 						switch_regex_safe_free(re);
 						switch_safe_free(field_expanded);
 						continue;
@@ -267,10 +267,11 @@ SWITCH_STANDARD_DIALPLAN(asterisk_dialplan_hunt)
 				}
 
 				if (strchr(expression, '(')) {
-					switch_perform_substitution(re, proceed, argument, field_data, substituted, sizeof(substituted), ovector);
+					switch_perform_substitution(match_data, argument, substituted, sizeof(substituted));
 					argument = substituted;
 				}
 
+				switch_regex_match_safe_free(match_data);
 				switch_regex_safe_free(re);
 
 				if (!extension) {
diff --git a/src/mod/dialplans/mod_dialplan_xml/mod_dialplan_xml.c b/src/mod/dialplans/mod_dialplan_xml/mod_dialplan_xml.c
index f85a985..cab5228 100644
--- a/src/mod/dialplans/mod_dialplan_xml/mod_dialplan_xml.c
+++ b/src/mod/dialplans/mod_dialplan_xml/mod_dialplan_xml.c
@@ -103,6 +103,7 @@ static int parse_exten(switch_core_session_t *session, switch_caller_profile_t *
 	int proceed = 0, save_proceed = 0;
 	char *expression_expanded = NULL, *field_expanded = NULL;
 	switch_regex_t *re = NULL, *save_re = NULL;
+	switch_regex_match_t *match_data = NULL, *save_match_data = NULL;
 	int offset = 0;
 	const char *tmp, *tzoff = NULL, *tzname_ = NULL, *req_nesta = NULL;
 	char nbuf[128] = "";
@@ -170,7 +171,6 @@ static int parse_exten(switch_core_session_t *session, switch_caller_profile_t *
 		char *expression = NULL, *save_expression = NULL, *save_field_data = NULL;
 		char *regex_rule = NULL;
 		const char *field_data = NULL;
-		int ovector[30];
 		switch_bool_t anti_action = SWITCH_TRUE;
 		break_t do_break_i = BREAK_ON_FALSE;
 		int time_match;
@@ -292,7 +292,7 @@ static int parse_exten(switch_core_session_t *session, switch_caller_profile_t *
 						field_data = "";
 					}
 
-					if ((proceed = switch_regex_perform(field_data, expression, &re, ovector, sizeof(ovector) / sizeof(ovector[0])))) {
+					if ((proceed = switch_regex_perform(field_data, expression, &re, &match_data))) {
 						if ( switch_core_test_flag(SCF_DIALPLAN_TIMESTAMPS) ) {
 							switch_log_printf(SWITCH_CHANNEL_SESSION_LOG(session), SWITCH_LOG_DEBUG,
 										  "%sDialplan: %s Regex (PASS) [%s] %s(%s) =~ /%s/ match=%s\n", space,
@@ -344,20 +344,24 @@ static int parse_exten(switch_core_session_t *session, switch_caller_profile_t *
 					switch_snprintf(var, sizeof(var), "DP_REGEX_MATCH_%d", total);
 
 					switch_channel_set_variable(channel, var, NULL);
-					switch_capture_regex(re, proceed, field_data, ovector, var, switch_regex_set_var_callback, session);
+					switch_capture_regex(match_data, proceed, var, switch_regex_set_var_callback, session);
 
 					switch_safe_free(save_expression);
 					switch_safe_free(save_field_data);
+					switch_regex_match_safe_free(save_match_data);
 					switch_regex_safe_free(save_re);
 
 					save_expression = strdup(expression);
 					save_field_data = strdup(field_data);
 					save_re = re;
+					save_match_data = match_data;
 					save_proceed = proceed;
 
 					re = NULL;
+					match_data = NULL;
 				}
 
+				switch_regex_match_safe_free(match_data);
 				switch_regex_safe_free(re);
 
 				switch_safe_free(field_expanded);
@@ -406,7 +410,7 @@ static int parse_exten(switch_core_session_t *session, switch_caller_profile_t *
 					field_data = "";
 				}
 
-				if ((proceed = switch_regex_perform(field_data, expression, &re, ovector, sizeof(ovector) / sizeof(ovector[0])))) {
+				if ((proceed = switch_regex_perform(field_data, expression, &re, &match_data))) {
 					if ( switch_core_test_flag(SCF_DIALPLAN_TIMESTAMPS) ) {
 						switch_log_printf(SWITCH_CHANNEL_SESSION_LOG(session), SWITCH_LOG_DEBUG,
 									  "%sDialplan: %s Regex (PASS) [%s] %s(%s) =~ /%s/ break=%s\n", space,
@@ -446,7 +450,9 @@ static int parse_exten(switch_core_session_t *session, switch_caller_profile_t *
 
 		if (save_re) {
 			re = save_re;
+			match_data = save_match_data;
 			save_re = NULL;
+			save_match_data = NULL;
 
 			expression = expression_expanded = save_expression;
 			save_expression = NULL;
@@ -506,7 +512,7 @@ static int parse_exten(switch_core_session_t *session, switch_caller_profile_t *
 		} else {
 			if (field && expression && strchr(expression, '(')) {
 				switch_channel_set_variable(channel, "DP_MATCH", NULL);
-				switch_capture_regex(re, proceed, field_data, ovector, "DP_MATCH", switch_regex_set_var_callback, session);
+				switch_capture_regex(match_data, proceed, "DP_MATCH", switch_regex_set_var_callback, session);
 			}
 
 			for (xaction = switch_xml_child(xcond, "action"); xaction; xaction = xaction->next) {
@@ -534,7 +540,7 @@ static int parse_exten(switch_core_session_t *session, switch_caller_profile_t *
 						goto done;
 					}
 					memset(substituted, 0, len);
-					switch_perform_substitution(re, proceed, data, field_data, substituted, len, ovector);
+					switch_perform_substitution(match_data, data, substituted, len);
 					app_data = substituted;
 				} else {
 					app_data = data;
@@ -571,6 +577,8 @@ static int parse_exten(switch_core_session_t *session, switch_caller_profile_t *
 				switch_safe_free(substituted);
 			}
 		}
+
+		switch_regex_match_safe_free(match_data);
 		switch_regex_safe_free(re);
 
 		if (((anti_action == SWITCH_FALSE && do_break_i == BREAK_ON_TRUE) ||
@@ -591,6 +599,7 @@ static int parse_exten(switch_core_session_t *session, switch_caller_profile_t *
 	}
 
   done:
+	switch_regex_match_safe_free(match_data);
 	switch_regex_safe_free(re);
 	switch_safe_free(field_expanded);
 	switch_safe_free(expression_expanded);
diff --git a/src/mod/endpoints/mod_sofia/sofia_glue.c b/src/mod/endpoints/mod_sofia/sofia_glue.c
index 1452a7c..33904aa 100644
--- a/src/mod/endpoints/mod_sofia/sofia_glue.c
+++ b/src/mod/endpoints/mod_sofia/sofia_glue.c
@@ -911,8 +911,6 @@ char *sofia_glue_get_extra_headers(switch_channel_t *channel, const char *prefix
 	switch_stream_handle_t stream = { 0 };
 	switch_event_header_t *hi = NULL;
 	const char *exclude_regex = NULL;
-	switch_regex_t *re = NULL;
-	int ovector[30] = {0};
 
 	exclude_regex = switch_channel_get_variable(channel, "exclude_outgoing_extra_header");
 	SWITCH_STANDARD_STREAM(stream);
@@ -926,12 +924,11 @@ char *sofia_glue_get_extra_headers(switch_channel_t *channel, const char *prefix
 			}
 
 			if (!strncasecmp(name, prefix, strlen(prefix))) {
-				if ( !exclude_regex || !(/*proceed*/ switch_regex_perform(name, exclude_regex, &re, ovector, sizeof(ovector) / sizeof(ovector[0])))) {
+				if ( !exclude_regex || !(/*proceed*/ switch_regex(name, exclude_regex))) {
 					const char *hname = name + strlen(prefix);
 					stream.write_function(&stream, "%s: %s\r\n", hname, value);
 				} else {
 					switch_log_printf(SWITCH_CHANNEL_LOG, SWITCH_LOG_DEBUG, "Ignoring Extra Header [%s] , matches exclude_outgoing_extra_header [%s]\n", name, exclude_regex);
-					switch_regex_safe_free(re);
 				}
 			}
 		}
diff --git a/src/mod/event_handlers/mod_event_socket/mod_event_socket.c b/src/mod/event_handlers/mod_event_socket/mod_event_socket.c
index 935f726..1c35b6f 100644
--- a/src/mod/event_handlers/mod_event_socket/mod_event_socket.c
+++ b/src/mod/event_handlers/mod_event_socket/mod_event_socket.c
@@ -349,10 +349,7 @@ static void event_handler(switch_event_t *event)
 						}
 
 						if (*hp->value == '/') {
-							switch_regex_t *re = NULL;
-							int ovector[30];
-							cmp = !!switch_regex_perform(hval, comp_to, &re, ovector, sizeof(ovector) / sizeof(ovector[0]));
-							switch_regex_safe_free(re);
+							cmp = !!switch_regex(hval, comp_to);
 						} else {
 							cmp = !strcasecmp(hval, comp_to);
 						}
diff --git a/src/switch_channel.c b/src/switch_channel.c
index 9c7b8e4..6866f35 100644
--- a/src/switch_channel.c
+++ b/src/switch_channel.c
@@ -33,7 +33,8 @@
 
 #include <switch.h>
 #include <switch_channel.h>
-#include <pcre.h>
+#define PCRE2_CODE_UNIT_WIDTH 8
+#include <pcre2.h>
 
 struct switch_cause_table {
 	const char *name;
@@ -4548,21 +4549,22 @@ SWITCH_DECLARE(switch_status_t) switch_channel_set_timestamps(switch_channel_t *
 		char *digit_string = dtstr;
 		char *X = NULL;
 		switch_regex_t *re = NULL;
+		switch_regex_match_t *match_data = NULL;
 		char *substituted = NULL;
 
 		if (!zstr(var)) {
 			int proceed = 0;
-			int ovector[30];
 
-			if ((proceed = switch_regex_perform(dtstr, var, &re, ovector, sizeof(ovector) / sizeof(ovector[0])))) {
+			if ((proceed = switch_regex_perform(dtstr, var, &re, &match_data))) {
 				int len = (strlen(dtstr) + strlen(var) + 10) * proceed;
 				int i = 0;
 				const char *replace = NULL;
+				PCRE2_SIZE replace_size;
 
 				X = malloc(len);
 
 				for (i = 0; i < proceed; i++) {
-					if (pcre_get_substring(dtstr, ovector, proceed, i, &replace) >= 0) {
+					if (pcre2_substring_get_bynumber(match_data, i, (PCRE2_UCHAR **)&replace, &replace_size) >= 0) {
 						if (replace) {
 							switch_size_t plen = strlen(replace);
 							memset(X, 'X', plen);
@@ -4571,7 +4573,7 @@ SWITCH_DECLARE(switch_status_t) switch_channel_set_timestamps(switch_channel_t *
 							switch_safe_free(substituted);
 							substituted = switch_string_replace(substituted ? substituted : dtstr, replace, X);
 							
-							pcre_free_substring(replace);
+							pcre2_substring_free((PCRE2_UCHAR *)replace);
 						}
 					}
 				}
@@ -4583,6 +4585,7 @@ SWITCH_DECLARE(switch_status_t) switch_channel_set_timestamps(switch_channel_t *
 		}
 
 		switch_channel_set_variable(channel, "digits_dialed", digit_string);
+		switch_regex_match_safe_free(match_data);
 		switch_regex_safe_free(re);
 		switch_safe_free(substituted);
 		switch_safe_free(X);
diff --git a/src/switch_ivr.c b/src/switch_ivr.c
index 1443f98..098f7a8 100644
--- a/src/switch_ivr.c
+++ b/src/switch_ivr.c
@@ -4343,8 +4343,6 @@ SWITCH_DECLARE(char *) switch_ivr_check_presence_mapping(const char *exten_name,
 	switch_xml_t cfg, xml, x_domains, x_domain, x_exten;
 	char *r = NULL;
 	switch_event_t *params = NULL;
-	switch_regex_t *re = NULL;
-	int proceed = 0, ovector[100];
 
 	switch_event_create(&params, SWITCH_EVENT_REQUEST_PARAMS);
 	switch_assert(params);
@@ -4375,17 +4373,11 @@ SWITCH_DECLARE(char *) switch_ivr_check_presence_mapping(const char *exten_name,
 			const char *regex = switch_xml_attr(x_exten, "regex");
 			const char *proto = switch_xml_attr(x_exten, "proto");
 
-			if (!zstr(regex) && !zstr(proto)) {
-				proceed = switch_regex_perform(exten_name, regex, &re, ovector, sizeof(ovector) / sizeof(ovector[0]));
-				switch_regex_safe_free(re);
-
-				if (proceed) {
-					switch_log_printf(SWITCH_CHANNEL_LOG, SWITCH_LOG_DEBUG1, "Mapping %s@%s to proto %s matching expression [%s]\n",
-									  exten_name, domain_name, proto, regex);
-					r = strdup(proto);
-					goto end;
-				}
-
+			if (!zstr(regex) && !zstr(proto) && switch_regex(exten_name, regex)) {
+				switch_log_printf(SWITCH_CHANNEL_LOG, SWITCH_LOG_DEBUG1, "Mapping %s@%s to proto %s matching expression [%s]\n",
+								  exten_name, domain_name, proto, regex);
+				r = strdup(proto);
+				goto end;
 			}
 		}
 	}
diff --git a/src/switch_ivr_async.c b/src/switch_ivr_async.c
index 4075f0a..ba80ea1 100644
--- a/src/switch_ivr_async.c
+++ b/src/switch_ivr_async.c
@@ -370,12 +370,11 @@ static dm_match_t switch_ivr_dmachine_check_match(switch_ivr_dmachine_t *dmachin
 	for(bp = dmachine->realm->binding_list; bp; bp = bp->next) {
 		if (bp->is_regex) {
 			if (bp->repl) {
-				int ovector[30] = { 0 };
 				int proceed = 0;
 				switch_regex_t *re = NULL;
-
+				switch_regex_match_t *match_data = NULL;
 				
-				proceed = switch_regex_perform(dmachine->digits, bp->digits, &re, ovector, sizeof(ovector) / sizeof(ovector[0]));
+				proceed = switch_regex_perform(dmachine->digits, bp->digits, &re, &match_data);
 				
 				if (proceed) {
 					char *substituted = NULL;
@@ -385,12 +384,14 @@ static dm_match_t switch_ivr_dmachine_check_match(switch_ivr_dmachine_t *dmachin
 					substituted = malloc(len);
 					switch_assert(substituted);
 					memset(substituted, 0, len);
-					switch_perform_substitution(re, proceed, bp->repl, dmachine->digits, substituted, len, ovector);
+					switch_perform_substitution(match_data, bp->repl, substituted, len);
 
 					if (!bp->substituted || strcmp(substituted, bp->substituted)) {
 						bp->substituted = switch_core_strdup(dmachine->pool, substituted);
 					}
+
 					free(substituted);
+					switch_regex_match_safe_free(match_data);
 					switch_regex_safe_free(re);
 					bp->rmatch = 1;
 				} else {
diff --git a/src/switch_ivr_menu.c b/src/switch_ivr_menu.c
index 255c9a4..39fc539 100644
--- a/src/switch_ivr_menu.c
+++ b/src/switch_ivr_menu.c
@@ -553,14 +553,15 @@ SWITCH_DECLARE(switch_status_t) switch_ivr_menu_execute(switch_core_session_t *s
 
 				if (ap->re) {
 					switch_regex_t *re = NULL;
-					int ovector[30];
+					switch_regex_match_t *match_data = NULL;
 
-					if ((ok = switch_regex_perform(menu->buf, ap->bind, &re, ovector, sizeof(ovector) / sizeof(ovector[0])))) {
-						switch_perform_substitution(re, ok, ap->arg, menu->buf, substituted, sizeof(substituted), ovector);
+					if ((ok = switch_regex_perform(menu->buf, ap->bind, &re, &match_data))) {
+						switch_perform_substitution(match_data, ap->arg, substituted, sizeof(substituted));
 						use_arg = substituted;
 					}
 					switch_log_printf(SWITCH_CHANNEL_SESSION_LOG(session), SWITCH_LOG_DEBUG, "action regex [%s] [%s] [%d]\n", menu->buf, ap->bind, ok);
 
+					switch_regex_match_safe_free(match_data);
 					switch_regex_safe_free(re);
 				} else {
 					ok = !strcmp(menu->buf, ap->bind);
diff --git a/src/switch_ivr_play_say.c b/src/switch_ivr_play_say.c
index b39d42f..5d926f1 100644
--- a/src/switch_ivr_play_say.c
+++ b/src/switch_ivr_play_say.c
@@ -178,7 +178,8 @@ SWITCH_DECLARE(switch_status_t) switch_ivr_phrase_macro_event(switch_core_sessio
 		char *field_expanded = NULL;
 		char *field_expanded_alloc = NULL;
 		switch_regex_t *re = NULL;
-		int proceed = 0, ovector[100];
+		switch_regex_match_t *match_data = NULL;
+		int proceed = 0;
 		switch_xml_t match = NULL;
 
 		searched = 1;
@@ -204,7 +205,7 @@ SWITCH_DECLARE(switch_status_t) switch_ivr_phrase_macro_event(switch_core_sessio
 
 		status = SWITCH_STATUS_SUCCESS;
 
-		if ((proceed = switch_regex_perform(field_expanded, pattern, &re, ovector, sizeof(ovector) / sizeof(ovector[0])))) {
+		if ((proceed = switch_regex_perform(field_expanded, pattern, &re, &match_data))) {
 			match = switch_xml_child(input, "match");
 		} else {
 			match = switch_xml_child(input, "nomatch");
@@ -224,12 +225,13 @@ SWITCH_DECLARE(switch_status_t) switch_ivr_phrase_macro_event(switch_core_sessio
 					len = (uint32_t) (strlen(data) + strlen(adata) + 10) * proceed;
 					if (!(substituted = malloc(len))) {
 						switch_log_printf(SWITCH_CHANNEL_SESSION_LOG(session), SWITCH_LOG_ERROR, "Memory Error!\n");
+						switch_regex_match_safe_free(match_data);
 						switch_regex_safe_free(re);
 						switch_safe_free(field_expanded_alloc);
 						goto done;
 					}
 					memset(substituted, 0, len);
-					switch_perform_substitution(re, proceed, adata, field_expanded, substituted, len, ovector);
+					switch_perform_substitution(match_data, adata, substituted, len);
 					odata = substituted;
 				} else {
 					odata = adata;
@@ -326,6 +328,7 @@ SWITCH_DECLARE(switch_status_t) switch_ivr_phrase_macro_event(switch_core_sessio
 			}
 		}
 
+		switch_regex_match_safe_free(match_data);
 		switch_regex_safe_free(re);
 		switch_safe_free(field_expanded_alloc);
 
diff --git a/src/switch_regex.c b/src/switch_regex.c
index 21e54b2..39fcd34 100644
--- a/src/switch_regex.c
+++ b/src/switch_regex.c
@@ -24,39 +24,47 @@
  * Contributor(s):
  *
  * Michael Jerris <mike@jerris.com>
+ * Christian Marangi <ansuelsmth@gmail.com> # PCRE2 conversion
  *
  *
- * switch_regex.c -- PCRE wrapper
+ * switch_regex.c -- PCRE2 wrapper
  *
  */
 
 #include <switch.h>
-#include <pcre.h>
+#define PCRE2_CODE_UNIT_WIDTH 8
+#include <pcre2.h>
 
 SWITCH_DECLARE(switch_regex_t *) switch_regex_compile(const char *pattern,
-													  int options, const char **errorptr, int *erroroffset, const unsigned char *tables)
+													  int options, int *errorcode, unsigned int *erroroffset, switch_regex_compile_context_t *ccontext)
 {
 
-	return (switch_regex_t *)pcre_compile(pattern, options, errorptr, erroroffset, tables);
+	return (switch_regex_t *)pcre2_compile((PCRE2_SPTR)pattern, PCRE2_ZERO_TERMINATED, options, errorcode, (PCRE2_SIZE *)erroroffset, ccontext);
+}
 
+SWITCH_DECLARE(int) switch_regex_copy_substring(switch_regex_match_t *match_data, int stringnumber, char *buffer, size_t *size)
+{
+	return pcre2_substring_copy_bynumber(match_data, stringnumber, (PCRE2_UCHAR *)buffer, (PCRE2_SIZE *)size);
 }
 
-SWITCH_DECLARE(int) switch_regex_copy_substring(const char *subject, int *ovector, int stringcount, int stringnumber, char *buffer, int size)
+SWITCH_DECLARE(void) switch_regex_match_free(void *data)
 {
-	return pcre_copy_substring(subject, ovector, stringcount, stringnumber, buffer, size);
+	pcre2_match_data_free(data);
 }
 
 SWITCH_DECLARE(void) switch_regex_free(void *data)
 {
-	pcre_free(data);
+	pcre2_code_free(data);
 
 }
 
-SWITCH_DECLARE(int) switch_regex_perform(const char *field, const char *expression, switch_regex_t **new_re, int *ovector, uint32_t olen)
+SWITCH_DECLARE(int) switch_regex_perform(const char *field, const char *expression, switch_regex_t **new_re, switch_regex_match_t **new_match_data)
 {
-	const char *error = NULL;
-	int erroffset = 0;
-	pcre *re = NULL;
+	int error_code = 0;
+	PCRE2_UCHAR error_str[128];
+	PCRE2_SIZE error_offset = 0;
+	pcre2_code *re = NULL;
+	pcre2_match_data *match_data;
 	int match_count = 0;
 	char *tmp = NULL;
 	uint32_t flags = 0;
@@ -87,52 +95,68 @@ SWITCH_DECLARE(int) switch_regex_perform(const char *field, const char *expressi
 		expression = tmp;
 		if (*opts) {
 			if (strchr(opts, 'i')) {
-				flags |= PCRE_CASELESS;
+				flags |= PCRE2_CASELESS;
 			}
 			if (strchr(opts, 's')) {
-				flags |= PCRE_DOTALL;
+				flags |= PCRE2_DOTALL;
 			}
 		}
 	}
 
-	re = pcre_compile(expression,	/* the pattern */
+	re = pcre2_compile((PCRE2_SPTR)expression,	/* the pattern */
+					  PCRE2_ZERO_TERMINATED,
 					  flags,	/* default options */
-					  &error,	/* for error message */
-					  &erroffset,	/* for error offset */
+					  &error_code,	/* for error code */
+					  &error_offset,	/* for error offset */
 					  NULL);	/* use default character tables */
-	if (error) {
-		switch_log_printf(SWITCH_CHANNEL_LOG, SWITCH_LOG_ERROR, "COMPILE ERROR: %d [%s][%s]\n", erroffset, error, expression);
-		switch_regex_safe_free(re);
+	if (!re) {
+		pcre2_get_error_message(error_code, error_str, 128);
+		switch_log_printf(SWITCH_CHANNEL_LOG, SWITCH_LOG_ERROR, "COMPILE ERROR: %zu [%s][%s]\n", error_offset, error_str, expression);
 		goto end;
 	}
 
-	match_count = pcre_exec(re,	/* result of pcre_compile() */
-							NULL,	/* we didn't study the pattern */
-							field,	/* the subject string */
+	match_data = pcre2_match_data_create_from_pattern(re, NULL);
+
+	match_count = pcre2_match(re,	/* result of pcre_compile() */
+							(PCRE2_SPTR)field,	/* the subject string */
 							(int) strlen(field),	/* the length of the subject string */
 							0,	/* start at offset 0 in the subject */
 							0,	/* default options */
-							ovector,	/* vector of integers for substring information */
-							olen);	/* number of elements (NOT size in bytes) */
+							match_data,	/* vector of integers for substring information */
+							NULL);	/* number of elements (NOT size in bytes) */
 
 
 	if (match_count <= 0) {
+		switch_regex_match_safe_free(match_data);
 		switch_regex_safe_free(re);
 		match_count = 0;
 	}
 
-	*new_re = (switch_regex_t *) re;
+	if (!new_re || !new_match_data) {
+		switch_regex_match_safe_free(match_data);
+		switch_regex_safe_free(re);
+	}
+
+	if (new_re) {
+		*new_re = (switch_regex_t *)re;
+	}
+
+	if (new_match_data) {
+		*new_match_data = (switch_regex_match_t *)match_data;
+	}
 
   end:
 	switch_safe_free(tmp);
+
 	return match_count;
 }
 
-SWITCH_DECLARE(void) switch_perform_substitution(switch_regex_t *re, int match_count, const char *data, const char *field_data,
-												 char *substituted, switch_size_t len, int *ovector)
+SWITCH_DECLARE(void) switch_perform_substitution(switch_regex_match_t *match_data, const char *data,
+												 char *substituted, switch_size_t len)
 {
 	char index[10] = "";
 	const char *replace = NULL;
+	PCRE2_SIZE replace_size;
 	switch_size_t x, y = 0, z = 0;
 	int num = 0;
 	int brace;
@@ -174,14 +198,15 @@ SWITCH_DECLARE(void) switch_perform_substitution(switch_regex_t *re, int match_c
 				num = -1;
 			}
 
-			if (pcre_get_substring(field_data, ovector, match_count, num, &replace) >= 0) {
+			if (pcre2_substring_get_bynumber(match_data, num, (PCRE2_UCHAR **)&replace, &replace_size) >= 0) {
 				if (replace) {
 					switch_size_t r;
 
 					for (r = 0; r < strlen(replace) && y < (len - 1); r++) {
 						substituted[y++] = replace[r];
 					}
-					pcre_free_substring(replace);
+
+					pcre2_substring_free((PCRE2_UCHAR *)replace);
 				}
 			}
 		} else {
@@ -193,20 +218,21 @@ SWITCH_DECLARE(void) switch_perform_substitution(switch_regex_t *re, int match_c
 }
 
 
-SWITCH_DECLARE(void) switch_capture_regex(switch_regex_t *re, int match_count, const char *field_data,
-										  int *ovector, const char *var, switch_cap_callback_t callback, void *user_data)
+SWITCH_DECLARE(void) switch_capture_regex(switch_regex_match_t *match_data, int match_count,
+										  const char *var, switch_cap_callback_t callback, void *user_data)
 
 {
 
 
 	const char *replace;
+	PCRE2_SIZE replace_size;
 	int i;
 
 	for (i = 0; i < match_count; i++) {
-		if (pcre_get_substring(field_data, ovector, match_count, i, &replace) >= 0) {
+		if (pcre2_substring_get_bynumber(match_data, i, (PCRE2_UCHAR **)&replace, &replace_size) >= 0) {
 			if (replace) {
-				callback(var, replace, user_data);
-				pcre_free_substring(replace);
+				callback(var, (const char *)replace, user_data);
+				pcre2_substring_free((PCRE2_UCHAR *)replace);
 			}
 		}
 	}
@@ -214,12 +240,13 @@ SWITCH_DECLARE(void) switch_capture_regex(switch_regex_t *re, int match_count, c
 
 SWITCH_DECLARE(switch_status_t) switch_regex_match_partial(const char *target, const char *expression, int *partial)
 {
-	const char *error = NULL;	/* Used to hold any errors                                           */
-	int error_offset = 0;		/* Holds the offset of an error                                      */
-	pcre *pcre_prepared = NULL;	/* Holds the compiled regex                                          */
+	PCRE2_UCHAR error[128]; /* Used to hold any errors                                           */
+	int error_code = 0;	/* Holds the code of an error                                           */
+	PCRE2_SIZE error_offset = 0;		/* Holds the offset of an error                                      */
+	pcre2_code *pcre_prepared = NULL;	/* Holds the compiled regex                                          */
 	int match_count = 0;		/* Number of times the regex was matched                             */
-	int offset_vectors[255];	/* not used, but has to exist or pcre won't even try to find a match */
-	int pcre_flags = 0;
+	pcre2_match_data *match_data;
+	int pcre2_flags = 0;
 	uint32_t flags = 0;
 	char *tmp = NULL;
 	switch_status_t status = SWITCH_STATUS_FALSE;
@@ -239,43 +266,44 @@ SWITCH_DECLARE(switch_status_t) switch_regex_match_partial(const char *target, c
 		expression = tmp;
 		if (*opts) {
 			if (strchr(opts, 'i')) {
-				flags |= PCRE_CASELESS;
+				flags |= PCRE2_CASELESS;
 			}
 			if (strchr(opts, 's')) {
-				flags |= PCRE_DOTALL;
+				flags |= PCRE2_DOTALL;
 			}
 		}
 	}
 
 	/* Compile the expression */
-	pcre_prepared = pcre_compile(expression, flags, &error, &error_offset, NULL);
+	pcre_prepared = pcre2_compile((PCRE2_SPTR)expression, PCRE2_ZERO_TERMINATED, flags, &error_code, &error_offset, NULL);
 
 	/* See if there was an error in the expression */
-	if (error != NULL) {
-		/* Clean up after ourselves */
-		if (pcre_prepared) {
-			pcre_free(pcre_prepared);
-			pcre_prepared = NULL;
-		}
+	if (!pcre_prepared) {
+		pcre2_get_error_message(error_code, error, 128);
+
 		/* Note our error */
 		switch_log_printf(SWITCH_CHANNEL_LOG, SWITCH_LOG_ERROR,
-						  "Regular Expression Error expression[%s] error[%s] location[%d]\n", expression, error, error_offset);
+						  "Regular Expression Error expression[%s] error[%s] location[%zu]\n", expression, error, error_offset);
 
 		/* We definitely didn't match anything */
 		goto end;
 	}
 
 	if (*partial) {
-		pcre_flags = PCRE_PARTIAL;
+		pcre2_flags = PCRE2_PARTIAL_SOFT;
 	}
 
 	/* So far so good, run the regex */
+	match_data = pcre2_match_data_create_from_pattern(pcre_prepared, NULL);
+
 	match_count =
-		pcre_exec(pcre_prepared, NULL, target, (int) strlen(target), 0, pcre_flags, offset_vectors, sizeof(offset_vectors) / sizeof(offset_vectors[0]));
+		pcre2_match(pcre_prepared, (PCRE2_SPTR)target, (int) strlen(target), 0, pcre2_flags, match_data, NULL);
+
+	pcre2_match_data_free(match_data);
 
 	/* Clean up */
 	if (pcre_prepared) {
-		pcre_free(pcre_prepared);
+		pcre2_code_free(pcre_prepared);
 		pcre_prepared = NULL;
 	}
 
@@ -285,7 +313,7 @@ SWITCH_DECLARE(switch_status_t) switch_regex_match_partial(const char *target, c
 	if (match_count > 0) {
 		*partial = 0;
 		switch_goto_status(SWITCH_STATUS_SUCCESS, end);
-	} else if (match_count == PCRE_ERROR_PARTIAL || match_count == PCRE_ERROR_BADPARTIAL) {
+	} else if (match_count == PCRE2_ERROR_PARTIAL) {
 		/* yes it is already set, but the code is clearer this way */
 		*partial = 1;
 		switch_goto_status(SWITCH_STATUS_SUCCESS, end);
diff --git a/src/switch_utils.c b/src/switch_utils.c
index aa3fc74..e4eabca 100644
--- a/src/switch_utils.c
+++ b/src/switch_utils.c
@@ -2081,8 +2081,9 @@ SWITCH_DECLARE(switch_status_t) switch_find_interface_ip(char *buf, int len, int
 SWITCH_DECLARE(switch_time_t) switch_str_time(const char *in)
 {
 	switch_time_exp_t tm = { 0 }, local_tm = { 0 };
-	int proceed = 0, ovector[30], time_only = 0;
+	int proceed = 0, time_only = 0;
 	switch_regex_t *re = NULL;
+	switch_regex_match_t *match_data = NULL;
 	char replace[1024] = "";
 	switch_time_t ret = 0, local_time = 0;
 	char *pattern = "^(\\d+)-(\\d+)-(\\d+)\\s*(\\d*):{0,1}(\\d*):{0,1}(\\d*)";
@@ -2092,66 +2093,78 @@ SWITCH_DECLARE(switch_time_t) switch_str_time(const char *in)
 	switch_time_exp_lt(&tm, switch_micro_time_now());
 
 
-	if ((time_only = switch_regex_perform(in, pattern3, &re, ovector, sizeof(ovector) / sizeof(ovector[0])))) {
+	if ((time_only = switch_regex_perform(in, pattern3, &re, &match_data))) {
 		tm.tm_hour = 0;
 		tm.tm_min = 0;
 		tm.tm_sec = 0;
 	} else {
 		tm.tm_year = tm.tm_mon = tm.tm_mday = tm.tm_hour = tm.tm_min = tm.tm_sec = tm.tm_usec = 0;
 
-		if (!(proceed = switch_regex_perform(in, pattern, &re, ovector, sizeof(ovector) / sizeof(ovector[0])))) {
+		if (!(proceed = switch_regex_perform(in, pattern, &re, &match_data))) {
+			switch_regex_match_safe_free(match_data);
 			switch_regex_safe_free(re);
-			proceed = switch_regex_perform(in, pattern2, &re, ovector, sizeof(ovector) / sizeof(ovector[0]));
+			proceed = switch_regex_perform(in, pattern2, &re, &match_data);
 		}
 	}
 
 	if (proceed || time_only) {
+		size_t replace_size;
 
 		if (time_only > 1) {
-			switch_regex_copy_substring(in, ovector, time_only, 1, replace, sizeof(replace));
+			replace_size = sizeof(replace);
+			switch_regex_copy_substring(match_data, 1, replace, &replace_size);
 			tm.tm_hour = atoi(replace);
 		}
 
 		if (time_only > 2) {
-			switch_regex_copy_substring(in, ovector, time_only, 2, replace, sizeof(replace));
+			replace_size = sizeof(replace);
+			switch_regex_copy_substring(match_data, 2, replace, &replace_size);
 			tm.tm_min = atoi(replace);
 		}
 
 		if (time_only > 3) {
-			switch_regex_copy_substring(in, ovector, time_only, 3, replace, sizeof(replace));
+			replace_size = sizeof(replace);
+			switch_regex_copy_substring(match_data, 3, replace, &replace_size);
 			tm.tm_sec = atoi(replace);
 		}
 
 		if (proceed > 1) {
-			switch_regex_copy_substring(in, ovector, proceed, 1, replace, sizeof(replace));
+			replace_size = sizeof(replace);
+			switch_regex_copy_substring(match_data, 1, replace, &replace_size);
 			tm.tm_year = atoi(replace) - 1900;
 		}
 
 		if (proceed > 2) {
-			switch_regex_copy_substring(in, ovector, proceed, 2, replace, sizeof(replace));
+			replace_size = sizeof(replace);
+			switch_regex_copy_substring(match_data, 2, replace, &replace_size);
 			tm.tm_mon = atoi(replace) - 1;
 		}
 
 		if (proceed > 3) {
-			switch_regex_copy_substring(in, ovector, proceed, 3, replace, sizeof(replace));
+			replace_size = sizeof(replace);
+			switch_regex_copy_substring(match_data, 3, replace, &replace_size);
 			tm.tm_mday = atoi(replace);
 		}
 
 		if (proceed > 4) {
-			switch_regex_copy_substring(in, ovector, proceed, 4, replace, sizeof(replace));
+			replace_size = sizeof(replace);
+			switch_regex_copy_substring(match_data, 4, replace, &replace_size);
 			tm.tm_hour = atoi(replace);
 		}
 
 		if (proceed > 5) {
-			switch_regex_copy_substring(in, ovector, proceed, 5, replace, sizeof(replace));
+			replace_size = sizeof(replace);
+			switch_regex_copy_substring(match_data, 5, replace, &replace_size);
 			tm.tm_min = atoi(replace);
 		}
 
 		if (proceed > 6) {
-			switch_regex_copy_substring(in, ovector, proceed, 6, replace, sizeof(replace));
+			replace_size = sizeof(replace);
+			switch_regex_copy_substring(match_data, 6, replace, &replace_size);
 			tm.tm_sec = atoi(replace);
 		}
 		
+		switch_regex_match_safe_free(match_data);
 		switch_regex_safe_free(re);
 
 		switch_time_exp_get(&local_time, &tm);
@@ -2160,9 +2173,11 @@ SWITCH_DECLARE(switch_time_t) switch_str_time(const char *in)
 		tm.tm_gmtoff = local_tm.tm_gmtoff;
 
 		switch_time_exp_gmt_get(&ret, &tm);
+
 		return ret;
 	}
 
+	switch_regex_match_safe_free(match_data);
 	switch_regex_safe_free(re);
 
 	return ret;
